cranelift/codegen/src/ir/layout.rs

```diff
@@ -781,6 +781,86 @@ impl<'f> DoubleEndedIterator for Insts<'f> {
     }
 }
 
+#[cfg(feature = "enable-serde")]
+mod serde {
+    use ::serde::de::{Deserializer, Error, SeqAccess, Visitor};
+    use ::serde::ser::{SerializeSeq, Serializer};
+    use ::serde::{Deserialize, Serialize};
+    use core::convert::TryFrom;
+    use core::fmt;
+    use core::marker::PhantomData;
+
+    use super::*;
+
+    impl Serialize for Layout {
```

Comment:
Can you say a bit more about why we need a custom Serialize/Deserialize for `Layout`? This sort of code can be somewhat error-prone, especially when we update it later (with footguns such as silently forgetting to include a field we have added). At the very least, if we have to implement it this way, we should document the serialized sequence (it looks like a series of blocknum, inst_count, [insts] elements?) and add warning comments in `Layout` to note that only certain data is serialized.

Reply:
`Layout` uses linked lists of blocks and instructions. In addition each block and inst has a sequence number that monotonically increases. This wastes a lot of space. I replaced it with a list of blocks and block sizes followed by a list of instructions.

Reply:
OK, that's sort of what I expected -- makes sense. Could you add a detailed doc comment that describes the equivalent serialized data structure (list of elements of these types ...)?

Reply:
Done